director/profile.go

```diff
@@ -723,115 +738,174 @@ func VerifyMDMProfiles(profileListData types.ProfileListData, device types.Devic
+		if profileForVerification.HashedPayloadUUID == profileList.PayloadUUID && profileForVerification.PayloadIdentifier == profileList.PayloadIdentifier && profileForVerification.Installed == true {
+			return true, true
+		}
+
+		// Profile is present in profile list, profile should not be installed
+		if profileForVerification.PayloadIdentifier == profileList.PayloadIdentifier && profileForVerification.Installed == false {
+			return true, false
 		}
 	}
+	return false, false
+}
 
-	DeleteDeviceProfiles(devices, profilesToRemove)
+// Add to the appropriate install list
+func addProfileToInstallLists(profileForVerification ProfileForVerification, sharedProfilesToInstall []types.SharedProfile, profilesToInstall []types.DeviceProfile) ([]types.SharedProfile, []types.DeviceProfile) {
+	if profileForVerification.Type == "shared" {
+		var sharedProfile types.SharedProfile
+		sharedProfile.HashedPayloadUUID = profileForVerification.HashedPayloadUUID
+		sharedProfile.PayloadUUID = profileForVerification.PayloadUUID
+		sharedProfile.PayloadIdentifier = profileForVerification.PayloadIdentifier
+		sharedProfile.Installed = profileForVerification.Installed
+		sharedProfile.MobileconfigData = profileForVerification.MobileconfigData
+		sharedProfile.MobileconfigHash = profileForVerification.MobileconfigHash
+		sharedProfilesToInstall = append(sharedProfilesToInstall, sharedProfile)
+	}
 
-	err = db.DB.Model(&sharedProfile).Find(&sharedProfiles).Where("installed = false").Scan(&sharedProfiles).Error
-	if err != nil {
-		return errors.Wrap(err, "VerifyMDMProfiles: Cannot load shared profiles to remove")
+	if profileForVerification.Type == "device" {
+		var deviceProfile types.DeviceProfile
+		deviceProfile.HashedPayloadUUID = profileForVerification.HashedPayloadUUID
+		deviceProfile.PayloadUUID = profileForVerification.PayloadUUID
+		deviceProfile.PayloadIdentifier = profileForVerification.PayloadIdentifier
+		deviceProfile.Installed = profileForVerification.Installed
+		deviceProfile.MobileconfigData = profileForVerification.MobileconfigData
+		deviceProfile.MobileconfigHash = profileForVerification.MobileconfigHash
+		deviceProfile.DeviceUDID = profileForVerification.DeviceUDID
+		profilesToInstall = append(profilesToInstall, deviceProfile)
 	}
 
-	for i := range sharedProfiles {
-		savedSharedProfile := sharedProfiles[i]
-		for i := range profileListData.ProfileList {
-			incomingProfile := profileListData.ProfileList[i]
-			if savedSharedProfile.PayloadIdentifier == incomingProfile.PayloadIdentifier {
-				sharedProfilesToRemove = append(sharedProfilesToRemove, savedSharedProfile)
-			}
-		}
+	return sharedProfilesToInstall, profilesToInstall
+}
 
+// Add to the appropriate removal list
+func addProfileToRemovalLists(profileForVerification ProfileForVerification, sharedProfilesToRemove []types.SharedProfile, profilesToRemove []types.DeviceProfile) ([]types.SharedProfile, []types.DeviceProfile) {
```

Comment:
We had discussed dropping the concept of `types.SharedProfile` in favor or representing it in `types.DeviceProfile` with the device identifier as `*` or something. I'm guessing it's a complicated change so we'll do it in a future refactor?

Reply:
Yes, it would be a large refactor, and going this route makes it a smaller job.